selectors/views/custom_status.ts

```diff
@@ -7,18 +7,24 @@ import {getConfig} from 'mattermost-redux/selectors/entities/general';
 
 import {get} from 'mattermost-redux/selectors/entities/preferences';
 import {Preferences} from 'mattermost-redux/constants';
-import {UserCustomStatus} from 'mattermost-redux/types/users';
+import {CustomStatusDuration, UserCustomStatus} from 'mattermost-redux/types/users';
 
 import {GlobalState} from 'types/store';
-
-export function makeGetCustomStatus(): (state: GlobalState, userID?: string) => UserCustomStatus {
-    return createSelector(
-        (state: GlobalState, userID?: string) => (userID ? getUser(state, userID) : getCurrentUser(state)),
-        (user) => {
-            const userProps = user?.props || {};
-            return userProps.customStatus ? JSON.parse(userProps.customStatus) : undefined;
-        },
-    );
+import {getCurrentUserTimezone} from 'selectors/general';
+import {getCurrentDateTimeForTimezone} from 'utils/timezone';
+
+export function getCustomStatus(state: GlobalState, userID?: string): UserCustomStatus | undefined {
```

Comment:
So you're saying to make two separate selectors - one for custom status (memoized) and one for checking if custom status has expired (not memoized). But I don't see the benefit of making an extra selector if that's what you are saying because the amount of re-rendering will be same as the 2nd selector is not memoized and is using the same parameters. What we can do is pass customStatus in the 2nd selector like this -
```
export function isCustomStatusExpired(state: GlobalState, customStatus?: UserCustomStatus) {
    if (!customStatus || customStatus.duration === CustomStatusDuration.DONT_CLEAR) {
        return false;
    }
    
    const expiryTime = moment(customStatus.expires_at)
    const timezone = getCurrentUserTimezone(state);
    const currentTime = getCurrentMomentForTimezone(timezone)
    return currentTime.isSameOrAfter(expiryTime);
}
```

But doing this can increase the delay in expiring custom status which is already approx. 1 minute.

Reply:
I don't think that should change the delay. The only delay I see is in how long it takes `getCustomStatus` to be called after the status expires, and I'm pretty sure it would happen the same either way.

I'm not sure if I understand your suggestion, but the main thing I'm trying to solve is that `JSON.parse(userProps.customStatus)` creates a new object each time it's called, and I want to memoize that part of it. Since we need to memoize it per-caller, we'd still need a `makeGetCustomStatus`. I might have gotten ahead of myself in that suggestion though since I was also trying to ensure we didn't accidentally memoize a pre-expiry status after it had expired though.